File: src/main/java/com/milaboratory/mixcr/cli/postanalysis/CommandOverlapScatter.java
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli.postanalysis;

import com.milaboratory.miplots.ExportKt;
import com.milaboratory.miplots.stat.xcontinious.CorrelationMethod;
import com.milaboratory.mixcr.basictypes.Clone;
import com.milaboratory.mixcr.cli.CommonDescriptions;
import com.milaboratory.mixcr.cli.MiXCRCommand;
import com.milaboratory.mixcr.postanalysis.Dataset;
import com.milaboratory.mixcr.postanalysis.SetPreprocessor;
import com.milaboratory.mixcr.postanalysis.overlap.OverlapGroup;
import com.milaboratory.mixcr.postanalysis.overlap.OverlapUtil;
import com.milaboratory.mixcr.postanalysis.plots.OverlapScatter;
import com.milaboratory.mixcr.postanalysis.plots.OverlapScatterRow;
import com.milaboratory.mixcr.postanalysis.preproc.ElementPredicate;
import com.milaboratory.mixcr.postanalysis.preproc.OverlapPreprocessorAdapter;
import com.milaboratory.mixcr.postanalysis.ui.DownsamplingParameters;
import com.milaboratory.mixcr.util.OutputPortWithProgress;
import com.milaboratory.util.SmartProgressReporter;
import io.repseq.core.Chains;
import jetbrains.letsPlot.intern.Plot;
import org.jetbrains.kotlinx.dataframe.DataFrame;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static io.repseq.core.Chains.*;

@Command(name = "overlapScatterPlot",
        separator = " ",
        description = "Plot overlap scatter-plot.")
public class CommandOverlapScatter extends MiXCRCommand {
    @Parameters(description = "cloneset_1.{clns|clna}...", index = "0")
    public String in1;
    @Parameters(description = "cloneset_2.{clns|clna}...", index = "1")
    public String in2;
    @Parameters(description = "output.[pdf|eps|png|jpeg]", index = "2")
    public String out;

    @Option(description = "Chains to export",
            names = "--chains")
    public List<String> chains = null;

    @Option(description = CommonDescriptions.ONLY_PRODUCTIVE,
            names = {"--only-productive"})
    public boolean onlyProductive = false;

    @Option(description = CommonDescriptions.DOWNSAMPLING,
            names = {"--downsampling"},
            required = true)
    public String downsampling;

    @Option(description = CommonDescriptions.OVERLAP_CRITERIA,
            names = {"--criteria"})
    public String overlapCriteria = "CDR3|AA|V|J";

    @Option(description = "Correlation method to use. Possible value: pearson, kendal, spearman",
            names = {"--method"})
    public String method = "pearson";

    @Option(description = "Do not apply log10 to clonotype frequences",
            names = {"--no-log"})
    public boolean noLog;

    @Override
    protected List<String> getInputFiles() {
        return Arrays.asList(in1, in2);
    }

    @Override
    protected List<String> getOutputFiles() {
        return Collections.singletonList(out);
    }

    private static String fName(String file) {
        return Paths.get(file).toAbsolutePath().getFileName().toString();
    }

    private Path outputPath(NamedChains chains) {
        if (chains == Chains.ALL_NAMED)
            return Paths.get(out);
        String fName = fName(out);
        return Paths.get(out).toAbsolutePath().getParent().resolve(fName.substring(0, fName.length() - 3) + chains.name + ".pdf");
    }

    @Override
    public void run0() throws Exception {
        DownsamplingParameters preproc = DownsamplingParameters.
                parse(downsampling, CommandPa.extractTagsInfo(getInputFiles()), false, onlyProductive);

        for (NamedChains curChains : this.chains == null
                ? Arrays.asList(TRAD_NAMED, TRB_NAMED, TRG_NAMED, IGH_NAMED, IGK_NAMED, IGL_NAMED)
                : this.chains.stream().map(Chains::getNamedChains).collect(Collectors.toList())) {

            OverlapPreprocessorAdapter.Factory<Clone> downsampling = new OverlapPreprocessorAdapter.Factory<>(preproc.getPreproc(curChains.chains));

            Dataset<OverlapGroup<Clone>> dataset = SetPreprocessor.processDatasets(downsampling.newInstance(),
                    OverlapUtil.overlap(
                            Arrays.asList(in1, in2),
                            new ElementPredicate.IncludeChains(curChains.chains),
                            OverlapUtil.parseCriteria(overlapCriteria).ordering()))[0];

            try (OutputPortWithProgress<OverlapGroup<Clone>> port = dataset.mkElementsPort()) {
                SmartProgressReporter.startProgressReport("Processing " + curChains.name, port);
                DataFrame<OverlapScatterRow> df = OverlapScatter.INSTANCE.dataFrame(port);
                if (df.rowsCount() == 0) {
                    continue;
                }
                Plot plot = OverlapScatter.INSTANCE.plot(df,
                        new OverlapScatter.PlotParameters(
                                fName(in1),
                                fName(in2),
                                CorrelationMethod.Companion.parse(method),
                                !noLog)
                );
                ExportKt.writeFile(outputPath(curChains), plot);
            }
        }
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/postanalysis/CommandOverlapScatter.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli.postanalysis

import com.milaboratory.miplots.stat.xcontinious.CorrelationMethod
import com.milaboratory.miplots.writeFile
import com.milaboratory.mixcr.cli.ChainsUtil
import com.milaboratory.mixcr.cli.ChainsUtil.name
import com.milaboratory.mixcr.cli.ChainsUtil.toPath
import com.milaboratory.mixcr.cli.CommonDescriptions
import com.milaboratory.mixcr.cli.CommonDescriptions.Labels
import com.milaboratory.mixcr.cli.InputFileType
import com.milaboratory.mixcr.cli.MiXCRCommandWithOutputs
import com.milaboratory.mixcr.cli.ValidationException
import com.milaboratory.mixcr.postanalysis.SetPreprocessor
import com.milaboratory.mixcr.postanalysis.overlap.OverlapUtil
import com.milaboratory.mixcr.postanalysis.plots.OverlapScatter
import com.milaboratory.mixcr.postanalysis.plots.OverlapScatter.dataFrame
import com.milaboratory.mixcr.postanalysis.plots.OverlapScatter.plot
import com.milaboratory.mixcr.postanalysis.preproc.ChainsFilter
import com.milaboratory.mixcr.postanalysis.preproc.ElementPredicate.IncludeChains
import com.milaboratory.mixcr.postanalysis.preproc.OverlapPreprocessorAdapter
import com.milaboratory.mixcr.postanalysis.ui.DownsamplingParameters
import com.milaboratory.util.SmartProgressReporter
import picocli.CommandLine.Command
import picocli.CommandLine.Help.Visibility.ALWAYS
import picocli.CommandLine.Option
import picocli.CommandLine.Parameters
import java.nio.file.Path
import kotlin.io.path.nameWithoutExtension

@Command(description = ["Plot overlap scatter-plot."])
class CommandOverlapScatter : MiXCRCommandWithOutputs() {
    @Parameters(paramLabel = "cloneset_1.(clns|clna)", index = "0")
    lateinit var in1: Path

    @Parameters(paramLabel = "cloneset_2.(clns|clna)", index = "1")
    lateinit var in2: Path

    @Parameters(paramLabel = "output.${InputFileType.exportTypesLabel}", index = "2")
    lateinit var out: Path

    @Option(
        description = ["Chains to export."],
        names = ["--chains"],
        split = ",",
        paramLabel = Labels.CHAIN,
        order = OptionsOrder.main + 10_100
    )
    var chains: Set<String>? = null

    @Option(
        description = [CommonDescriptions.ONLY_PRODUCTIVE],
        names = ["--only-productive"],
        order = OptionsOrder.main + 10_200
    )
    var onlyProductive = false

    @Option(
        description = ["Choose ${CommonDescriptions.DOWNSAMPLING}"],
        names = ["--downsampling"],
        required = true,
        paramLabel = "(<type>|none)",
        order = OptionsOrder.main + 10_300
    )
    lateinit var downsampling: String

    @Option(
        description = [CommonDescriptions.OVERLAP_CRITERIA],
        names = ["--criteria"],
        showDefaultValue = ALWAYS,
        paramLabel = "<s>",
        order = OptionsOrder.main + 10_400
    )
    var overlapCriteria = "CDR3|AA|V|J"

    @Option(
        description = ["Correlation method to use. Possible value: \${COMPLETION-CANDIDATES}."],
        names = ["--method"],
        paramLabel = "<method>",
        showDefaultValue = ALWAYS,
        order = OptionsOrder.main + 10_500
    )
    var method: CorrelationMethod = CorrelationMethod.Pearson

    @Option(
        description = ["Do not apply log10 to clonotype frequencies."],
        names = ["--no-log"],
        order = OptionsOrder.main + 10_600
    )
    var noLog = false

    override val inputFiles
        get() = listOf(in1, in2)

    override val outputFiles
        get() = listOf(out)

    override fun validate() {
        ValidationException.requireFileType(in1, InputFileType.CLNX)
        ValidationException.requireFileType(in2, InputFileType.CLNX)
        ValidationException.requireFileType(out, InputFileType.exportTypes)
    }

    override fun run0() {
        val parameters = DownsamplingParameters.parse(
            downsampling,
            CommandPa.extractTagsInfo(inputFiles, !downsampling.equals("none", ignoreCase = true)),
            false, onlyProductive
        )

        var chainsToProcess = ChainsUtil.allChainsFromClnx(listOf(in1, in2))
        chainsToProcess = chains?.let { ChainsFilter.parseChainsList(it) } ?: chainsToProcess

        for (chain in chainsToProcess) {
            val downsampling = OverlapPreprocessorAdapter.Factory(parameters.getPreprocessor(chain))
            val dataset = SetPreprocessor.processDatasets(
                downsampling.newInstance(),
                OverlapUtil.overlap(
                    listOf(in1, in2).map { it.toString() },
                    IncludeChains(setOf(chain), false),
                    OverlapUtil.parseCriteria(overlapCriteria).ordering()
                )
            ).first()
            val plotParameters = OverlapScatter.PlotParameters(
                in1.nameWithoutExtension,
                in2.nameWithoutExtension,
                method,
                !noLog
            )
            dataset.mkElementsPort().use { port ->
                SmartProgressReporter.startProgressReport("Processing ${chain.name}", port)
                val df = dataFrame(port)
                if (df.rowsCount() == 0) return@use
                val plot = plot(df, plotParameters)
                writeFile(chain.toPath(out), plot)
            }
        }
    }
}
